crates/core_arch/src/riscv32/mod.rs

```diff
@@ -0,0 +1,481 @@
+//! RISC-V RV32 intrinsics
+
+use crate::arch::asm;
+
+/// Generates the `PAUSE` instruction
+///
+/// The PAUSE instruction is a HINT that indicates the current hart's rate of instruction retirement
+/// should be temporarily reduced or paused. The duration of its effect must be bounded and may be zero.
+#[inline]
+pub fn pause() {
+    unsafe { asm!(".insn i 0x0F, 0, x0, x0, 0x010", options(nomem, nostack)) }
+}
+
+/// Generates the `NOP` instruction
+///
+/// The NOP instruction does not change any architecturally visible state, except for
+/// advancing the `pc` and incrementing any applicable performance counters.
+#[inline]
+pub fn nop() {
+    unsafe { asm!("nop") }
```

Comment:
All of the `asm!` blocks here should use `nostack` since they don't push anything to the stack.

Additionally:
- Most should use `nomem`, except for loads which should use `readonly` and stores which should use neither.

Reply:
Thanks, I'll check it out.
There are many privileged memory fences as well, I'd also figure it out in these intrinsic functions.

Reply:
Good point about the fences, those should not use `nomem`/`readonly` because they need to act as a compiler fence.